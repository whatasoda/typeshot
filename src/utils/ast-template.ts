import type ts from 'typescript';
import { forEachChildDeep } from './ast';

interface MutableTemplateStringsArray extends Array<string> {
  raw: string[];
}

export abstract class AstTemplate<T, U extends any[]> {
  private isAccumulator: boolean = false;
  private isFreezed: boolean = false;
  protected template: MutableTemplateStringsArray;
  protected substitutions: T[];
  protected abstract getSubstitution(node: ts.Node): T | null;
  protected abstract serializeSubstitution(substitution: T, ...args: U): string;

  constructor(node: null, sourceText?: never);
  constructor(node: ts.Node, sourceText: string);
  constructor(node: ts.Node | null, sourceText: string = '') {
    const raw: string[] = [];
    this.template = Object.assign(raw, { raw });
    this.substitutions = [];
    if (!node) {
      this.isAccumulator = true;
      return;
    }

    let cursor = node.getStart();
    forEachChildDeep(node, (node) => {
      const substitution = this.getSubstitution(node);
      if (substitution === null) return;
      this.template.push(sourceText.slice(cursor, node.getStart()));
      this.substitutions.push(substitution);
      cursor = node.getEnd();
      return true;
    });

    this.template.push(sourceText.slice(cursor, node.getEnd()));
  }

  public serialize(...args: U): string {
    let acc = '';
    this.substitutions.forEach((substitution, i) => {
      acc += this.template[i];
      acc += this.serializeSubstitution(substitution, ...args);
    });
    acc += this.template[this.template.length - 1];
    return acc;
  }

  public append(templates: this[], separator: string = ''): this {
    if (!this.isAccumulator) throw new Error('DONT use AstTemplate.prototype.append with non-accumulator instance.');
    if (this.isFreezed) return this;
    const { template, substitutions } = this;
    const ignoreIndex = (template[0] = template[0] || '') === '' ? 0 : -1;
    templates.forEach(({ template: t, substitutions: s }, i) => {
      template[template.length - 1] += (i === ignoreIndex ? '' : separator) + t[0];
      template.push(...t.slice(1));
      substitutions.push(...s);
    });
    return this;
  }

  public wrap(start: string, end: string): this {
    if (this.isFreezed) return this;
    const { template: t } = this;
    t[0] = start + t[0];
    t[t.length - 1] += end;
    return this;
  }

  public ensureEnd(end: string): this {
    if (this.isFreezed) return this;
    const { template: t } = this;
    const last = t[t.length - 1];
    t[t.length - 1] = last.trimEnd().endsWith(end) ? last : last + end;
    return this;
  }

  public freeze() {
    this.isFreezed = true;
  }
}
