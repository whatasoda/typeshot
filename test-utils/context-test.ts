import { getContext } from '../src/program/context';

(getContext() as any).test();

throw new Error('test error');
