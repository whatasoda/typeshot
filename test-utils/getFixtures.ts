import ts from 'typescript';
import path from 'path';

export const FIXTURE_DIR = path.resolve(__dirname, '../__fixtures__');

export const getFixtures = jest.fn((name: string, setParentNodes: boolean) => {
  const fileName = `${FIXTURE_DIR}/${name}.ts`;
  const sourceText = ts.sys.readFile(fileName);
  if (!sourceText) throw new Error(`fixture file '${name}' is not found`);
  return ts.createSourceFile(fileName, sourceText, ts.ScriptTarget.Latest, setParentNodes, ts.ScriptKind.TS);
});
