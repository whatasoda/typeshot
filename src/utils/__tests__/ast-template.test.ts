import ts from 'typescript';
import { getFixtures } from '../../../test-utils/getFixtures';
import { AstTemplate } from '../ast-template';

const fixtrue = getFixtures('sample', true);

describe('AstTemplate', () => {
  class TestTemplate extends AstTemplate<string, []> {
    protected getSubstitution(node: ts.Node) {
      const text = node.getText();
      if (text === 'number') {
        return 'NUMBER';
      } else if (text === "'number'") {
        return "'NUMBER'";
      } else {
        return null;
      }
    }
    protected serializeSubstitution(substitution: string) {
      return substitution;
    }
  }

  const sourceText = fixtrue.getFullText();
  const setUp = () => {
    const t = new TestTemplate(fixtrue, sourceText);
    const { template, substitutions }: { template: TemplateStringsArray; substitutions: string[] } = t as any;
    const expectedBase = String.raw(template, ...substitutions);
    return { t, template, substitutions, expectedBase };
  };

  it('should create template successfully', () => {
    const { t, template, substitutions, expectedBase } = setUp();

    expect(`'${template.join(`',\n${'-'.repeat(10)}\n'`)}'`).toMatchSnapshot();
    expect(substitutions).toMatchSnapshot();
    expect(t.serialize()).toBe(expectedBase);
  });

  it('should serialize successfully', () => {
    const { t, expectedBase } = setUp();
    expect(t.serialize()).toBe(expectedBase);
  });

  it('should wrap template with received strings', () => {
    const { t, expectedBase } = setUp();
    t.wrap('---', '---');
    expect(t.serialize()).toBe(`---${expectedBase}---`);
  });

  it('should append templates', () => {
    const acc = new TestTemplate(null);
    const { t, expectedBase } = setUp();
    acc.append([t, t], '~~~').append([t], '@@@');
    expect(acc.serialize()).toBe(`${expectedBase}~~~${expectedBase}@@@${expectedBase}`);
  });

  it('throws error if trying to append to non-accumulator instance', () => {
    const { t } = setUp();
    expect(() => t.append([t], '')).toThrowError();
  });

  test('ensureEnd works correctly', () => {
    const { t, expectedBase } = setUp();
    t.ensureEnd('###');
    expect(t.serialize()).toBe(`${expectedBase}###`);
    t.ensureEnd('###');
    expect(t.serialize()).toBe(`${expectedBase}###`);
  });

  test('freeze prevents every operation except serialize', () => {
    const t = new TestTemplate(null);
    t.freeze();
    t.wrap('abcdefg', 'hijklmn');
    t.ensureEnd('opqrstu');
    t.append([t]);
    expect(t.serialize()).toBe('');
  });
});
