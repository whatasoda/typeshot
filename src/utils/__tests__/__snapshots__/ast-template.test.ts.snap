// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstTemplate should create template successfully 1`] = `
"'import typeshot from 'typeshot';

typeshot.openTrace\`
// GENERATED FILE -- DO NOT EDIT
\`;

interface TypeMap {
  text: string;
  ',
----------
': ',
----------
';
}

export interface FieldDefinition {
  id: string;
  type: keyof TypeMap;
  required: boolean;
}

typeshot.closeTrace();

const PropsType = typeshot.registerTypeDefinition((fields: FieldDefinition[], makeType) => {
  const acc = {};
  const resolvedIdSet = new Set<string>();

  fields.forEach(({ id, type, required }) => {
    if (resolvedIdSet.has(id)) throw new Error('Duplicated field id found!');

    const t = required
      ? makeType<{ [K in typeof id]-?: TypeMap[typeof type] } & string[]>({ id, type })
      : makeType<{ [K in typeof id]+?: TypeMap[typeof type] } & string[]>({ id, type });
    Object.assign(acc, t);
  });
  const aaa = 'aaa';
  (acc as any).aaa = makeType<',
----------
' | typeof aaa>({ aaa });

  return acc;
});

const Props = PropsType([
  { id: 'foo', type: ',
----------
', required: false },
  { id: 'bar', type: ',
----------
', required: true },
  { id: 'baz', type: 'text', required: false },
  { id: 'qux', type: 'text', required: true },
]);

typeshot.print\`
  export type Hoge = \${Props.literal()}
\`;

typeshot.openTrace();

// hogehoge
typeshot.closeTrace();
'"
`;

exports[`AstTemplate should create template successfully 2`] = `
Array [
  "NUMBER",
  "NUMBER",
  "NUMBER",
  "'NUMBER'",
  "'NUMBER'",
]
`;
